### ISSUE TITLE
Inconsistent behavior among log, log2, log2Ceil, and log10

### ISSUE THREAD

**Issue by brianlheim** [supercollider/supercollider#2517]
Noticed this while experimenting with test cases for #2514.

Executed:
`
a = [-inf, -2.0, -1, 0, 1, 2.0, inf, 0/0 /* == nan */];
a.log.postln;
a.log2.postln;
a.log10.postln;
`

Post:
`
[ nan, nan, nan, nan, -inf, -0.69314718055995, 0, 0.69314718055995, inf, nan ]
[ inf, 1, 0, -1, -inf, -1, 0, 1, inf, nan ]
[ nan, nan, nan, nan, -inf, -0.30102999566398, 0, 0.30102999566398, inf, nan ]
`

log2 should return nan for all input < 0.

In addition, log2Ceil is a member of `Integer` and not `SimpleNumber` like the other 3, which means that it doesn't work on Floats. In addition, .log2Ceil returns `32` for any non-positive input:

`[-100,-2,-1,0,1].collect(_.log2Ceil)` -> `[ 32, 32, 32, 32, 0 ]`

My suggestion would be to deprecate log2Ceil as it's just two standard functions squashed together. However, I see that it's used a bit in some important C++ source files. Would it be possible to deprecate it from SC and rewrite the two places it's used in .sc files, but keep the C++ inline as legacy?

**Comment by telephon** [comment]
There are many operators that are two or more standard functions squashed together. This makes sense where performance is important.

**Comment by telephon** [comment]
But on the other hand, `logToCeil`seems to be really broken …

**Comment by brianlheim** [comment]
Right, but there's nothing in the SC codebase that suggests this particular combination is of value compared to log2Floor, log2Trunc, log10Ceil, etc.

**Comment by telephon** [comment]
ok, you are right. On the other hand, it works well for values `> 0`. And someone may rely on it. Better fixing it than adding an incompatibility, it does no harm after all?

**Comment by telephon** [comment]
compare also those:

```
a = { DC.ar([-inf, -2.0, -1, 0, 1, 2.0, inf]) };


{ a.value.log.poll(0); 0.0 }.play;
{ a.value.log2.poll(0); 0.0  }.play;
{ a.value.log10.poll(0); 0.0  }.play;
```

Returning:

```
-> Synth('temp__6' : 1010)
UGen Array [0]: -310.874
UGen Array [1]: -398.904
UGen Array [2]: -399.597
UGen Array [3]: -88.0297
UGen Array [4]: 0
UGen Array [5]: 0.693147
UGen Array [6]: 88.7228
-> Synth('temp__7' : 1011)
UGen Array [0]: -448.496
UGen Array [1]: -575.496
UGen Array [2]: -576.496
UGen Array [3]: -127
UGen Array [4]: 0
UGen Array [5]: 1
UGen Array [6]: 128
-> Synth('temp__8' : 1012)
UGen Array [0]: -135.011
UGen Array [1]: -173.242
UGen Array [2]: -173.543
UGen Array [3]: -38.2308
UGen Array [4]: 0
UGen Array [5]: 0.30103
UGen Array [6]: 38.5318
```

Note that it may be entirely ok that sclang interprets some math operations differently than standard calculus. It follows the design of `APL` (or `J`, http://www.kevinalbrecht.com/code/joy-mirror/j00ovr.html) in this respect.

There are four possibilities to choose from:

1. throw errors
2. return "bad" values
3. do exactly what the C-API would do
4. find a good way to make it cover the whole input range

**Comment by brianlheim** [comment]
I think it is harmful to keep inconsistent code just because someone might rely on its being broken or inconsistent. It reduces the coherence of the language. Moreover, the error "message not understood: .log2Ceil" that would show in the case of incompatibility is unambiguous and easily remedied by a quick replace of `.log2Ceil` with `.log2.ceil` (or `.log2.ceil.asInteger` if appropriate).

I did notice that UGen behavior and I'm glad you brought it up. I think nonstandard math functions are okay (as with `UGen-pow`) but to me the current behavior (especially at 0) makes no sense. I like the idea of implementing a fix similar to .pow in which `UGen-log2` is essentially shorthand for `.abs.log2`. `DC.ar(0).log2` should probably output NaN though. That's just what makes sense to me, I'm interested to hear what others have to say—perhaps there is some interesting use for the way it works now.

**Comment by telephon** [comment]
yes, I agree that it is better not to be too "permissive". It would be just good to check what the options are for these cases. If we have `neg(neg(a) ** b)` for `pow` on a ugen, then the most consistent solution for the `log` methods would be the correct inverse of the equivalent `pow`.
